packages/astro/actions-module.template.mjs

```diff
@@ -0,0 +1,49 @@
+import { ActionError, callSafely } from 'astro:actions';
+
+function toActionProxy(actionCallback = {}, aggregatedPath = '/_actions/') {
+	return new Proxy(actionCallback, {
+		get(target, objKey) {
+			const path = aggregatedPath + objKey.toString();
+			if (objKey in target) {
+				return target[objKey];
+			}
+			async function action(param) {
+				if (import.meta.env.SSR) {
+					throw new ActionError({
+						code: 'BAD_REQUEST',
+						message:
+							'Action unexpectedly called on the server. If this error is unexpected, share your feedback on our RFC discussion: https://github.com/withastro/roadmap/pull/912',
+					});
+				}
+				const headers = new Headers();
+				headers.set('Accept', 'application/json');
+				let body = param;
+				if (!(body instanceof FormData)) {
+					body = param ? JSON.stringify(param) : undefined;
+					headers.set('Content-Type', 'application/json');
+				}
+				const res = await fetch(path, {
+					method: 'POST',
+					body,
+					headers,
+				});
+				if (!res.ok) {
+					throw await ActionError.fromResponse(res);
+				}
+				const json = await res.json();
+				return json;
+			}
+			action.toString = () => path;
+			action.safe = (input) => {
+				return callSafely(() => action(input));
+			};
+			action.safe[Symbol.for('astro:action:safe')] = true;
+			action.safe.toString = () => path;
+			// recurse to construct queries for nested object paths
+			// ex. actions.user.admins.auth()
+			return toActionProxy(action, path + '.');
```

Comment:
Recursion! 😱 Can't we achieve the same using a loop?

Reply:
Hm, I can't see a way on first impression. We're building a URL with each dot chain `.` you add to a function. We can't predict how deep the function chaining will be, so we don't have an upper bound for a loop. To you have a suggestion if that's our constraint?

Reply:
> We can't predict how deep the function chaining will be, so we don't have an upper bound for a loop

That's unsettling :( no worries, we can review it later if that becomes an hot path